dask/array/utils.py

```diff
@@ -388,6 +393,36 @@ def zeros_like_safe(a, shape, **kwargs):
         return np.zeros(shape, **kwargs)
 
 
+def _array_like_safe(np_func, da_func, a, like, **kwargs):
+    if like is a and hasattr(a, "__array_function__"):
+        return a
+
+    if isinstance(like, Array):
+        return da_func(a, **kwargs)
+    elif isinstance(a, Array):
+        if _is_cupy_type(a._meta):
+            a = a.compute(scheduler="sync")
+
+    try:
+        return np_func(a, like=meta_from_array(like), **kwargs)
+    except TypeError:
+        return np_func(a, **kwargs)
+
+
+def array_safe(a, like, **kwargs):
+    """
+    If a is dask.array, return dask.array.asarray(a, **kwargs),
+    otherwise return np.asarray(a, like=like, **kwargs), dispatching
+    the call to the library that implements the like array. Note that
+    when a is a dask.Array but like isn't, this function will call
+    a.compute(scheduler="sync") before np.asarray, as downstream
+    libraries are unlikely to know how to convert a dask.Array.
```

Comment:
It looks like this comment could be updated as we'll only compute when `like` is a CuPy array

Reply:
Agreed, thanks for noticing that. I pushed that change in https://github.com/dask/dask/pull/7162/commits/433122a38eac7afec6f300802935cc5beaa8e953 .

Reply:
Marked as resolved. Though please let us know if there is anything else off in the correction